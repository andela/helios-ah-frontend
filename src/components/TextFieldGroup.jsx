import React from 'react';
import propTypes from 'prop-types';

const TextFieldGroup = ({
  field,
  value,
  error,
  type,
  onChange,
  onBlur,
  onInput,
  id,
  placeholder,
  errorFeedbackClass,
  validFeedbackClass
}) => {
  let checkValidity;
  let feedbackClass;
  if (error) feedbackClass = `${errorFeedbackClass} invalid-feedback`;
  else feedbackClass = `${validFeedbackClass}  valid-feedback`;
  if (error) checkValidity = 'form-control is-invalid';
  else if (value && !error) checkValidity = 'form-control is-valid';
  return (
    <div className="form-group">
      <input
        type={type}
        name={field}
        onInput={onInput}
        className={checkValidity}
        id={id}
        placeholder={placeholder}
        value={value}
        onChange={onChange}
        onBlur={onBlur}
      />
      <div className={feedbackClass}>
        {error && error}
      </div>
    </div>
  );
};

TextFieldGroup.propTypes = {
  field: propTypes.string.isRequired,
  value: propTypes.oneOfType([
    propTypes.string,
    propTypes.number,
  ]).isRequired,
  error: propTypes.string,
  type: propTypes.string,
  onChange: propTypes.func.isRequired,
  onInput: propTypes.func,
  id: propTypes.string,
  onBlur: propTypes.func,
  placeholder: propTypes.string,
  errorFeedbackClass: propTypes.string,
  validFeedbackClass: propTypes.string,
};

TextFieldGroup.defaultProps = {
  type: 'text',
  error: '',
  id: null,
  errorFeedbackClass: null,
  validFeedbackClass: null,
  onInput: null,
  onBlur: null,
  placeholder: ''
};

export default TextFieldGroup;
