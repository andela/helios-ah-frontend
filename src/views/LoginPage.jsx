import React from 'react';
import Links from '../components/Links';

const LoginPage = () => (
  <div>
    <Links />
    <h2>This is the Login Page</h2>
  </div>
);

export default LoginPage;
