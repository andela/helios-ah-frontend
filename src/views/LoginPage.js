import React, { Component } from 'react';
import TextFieldGroup from '../components/TextFieldGroup.jsx';
import frontendValidation from '../utils/frontendValidation';
import FlashMessagesList from '../components/FlashMessagesList.jsx'
import SubmitButton from '../components/SubmitButton.jsx';
import { Link } from 'react-router-dom';


class LoginPage extends Component {
  constructor(props) {
    super(props)
    this.state = {
      email: '',
      password: '',
      errors: {},
      isLoading: false
    }
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.onBlur = this.onBlur.bind(this);
    this.onInput = this.onInput.bind(this);
    this.onBlurError = {}
    this.onInputError = {}
  }

  onChange(e) {
    this.setState({
      [e.target.name]: e.target.value
    })
  }

  onInput(e) {
    const field = e.target.name;
    const value = e.target.value;
    this.setState({ [field]: value }, () => {
      const errors = frontendValidation(this.state)
      if (errors[field]) {
        this.onInputError[field] = errors[field]
        this.setState({ errors: this.onInputError[field] })
      } else {
        delete (this.onInputError[field])
      }
      this.setState({ errors: this.onInputError })
    })
  }

  isValid() {
    const errors = frontendValidation(this.state)
    if (Object.keys(errors).length > 0) {
      this.setState({ errors })
      return false
    }
    return true;
  }

  onBlur(e) {
    const field = e.target.name;
    const errors = frontendValidation(this.state)
    if (errors[field]) {
      this.onBlurError[field] = errors[field]
      this.setState({ errors: this.onBlurError[field] })
    } else {
      delete (this.onBlurError[field])
    }
    this.setState({ errors: this.onBlurError })
  }

  async onSubmit(e) {
    e.preventDefault();
    this.props.clearFlashMessages();
    if (this.isValid()) {
      this.setState({ errors: {}, isLoading: true })
      const loginResponse = await this.props.loginRequest(this.state)
      if (loginResponse) {
        this.setState({ isLoading: false })
        if (loginResponse.status === 200) {
          this.props.addFlashMessage({
            type: 'success',
            text: `${loginResponse.data.message}`
          })
        }
        else if (loginResponse.data.message === "Email or password does not exist") {
          this.props.addFlashMessage({
            type: 'error',
            text: 'Incorrect email or password'
          })
        }
        else {
          this.props.addFlashMessage({
            type: 'warning',
            text: `${loginResponse.data.message}`
          })
        }
      }
    }
  }

  render() {
    const { errors } = this.state
    return (
      <div className="container login-body">
        <div className="row">
          <div className="col-md-6">
            <span className="login-image-container">
            </span>
          </div>
          <div className="col-md-6 col-xs-12 right-login">
            <h3>WELCOME BACK.....</h3>
            <h5>LOGIN TO AUTHOR'S HAVEN</h5>

            <FlashMessagesList
              customAlertClass='custom-alert'
            />

            <form onSubmit={this.onSubmit} autoComplete="off">
              <TextFieldGroup
                error={errors && errors.email}
                placeholder="Email Address"
                field="email"
                onChange={this.onChange}
                onBlur={this.onBlur}
                value={this.state.email}
                errorFeedbackClass='login-invalid-feedback'
                validFeedbackClass='login-valid-feedback'
                onInput={this.onInput}
              />
              <TextFieldGroup
                error={errors && errors.password}
                placeholder="Password"
                field="password"
                type="password"
                onChange={this.onChange}
                onBlur={this.onBlur}
                value={this.state.password}
                errorFeedbackClass='login-invalid-feedback'
                validFeedbackClass='login-valid-feedback'
                onInput={this.onInput}
              />
              <div className="forgot-password">
                <Link to="/reset-password"><p>FORGOT PASSWORD?</p></Link>
              </div>
              <div className="login-buttons">
                <SubmitButton
                  buttonClass="login-btn"
                  isRequestSent={this.state.isLoading}
                  buttonValue="LOGIN"
                  columnAttribute="login-col"
                />
                <div className='signup-col'>
                  <div className="form-group text-center">
                    <Link to='/signup' className="signup-btn">
                      SIGNUP
                </Link>
                  </div>
                </div>
              </div>
              <div className="social-media-icons">
                <span className="facebook"></span>
                <span className="twitter"></span>
                <span className="google"></span>
              </div>
            </form>
          </div>
        </div >
      </div >
    );
  }
}

export default LoginPage;
