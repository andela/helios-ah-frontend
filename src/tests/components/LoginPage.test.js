import React from 'React';
import { shallow, mount } from 'enzyme';
import LoginPage from '../../views/LoginPage';
import TextFieldGroup from '../../components/TextFieldGroup';
import frontEndValidation from '../../utils/frontendValidation';

describe('Render Login Page', () => {
  let wrapper;

  beforeAll(() => {
    wrapper = shallow(<LoginPage />)
  });
  describe('render Login Page', () => {
    it('renders the login page component', () => {
      expect(wrapper.find('div').length).toEqual(7);
      expect(wrapper.find('div').find('span').length).toEqual(4);
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.find('h3').length).toEqual(1);
      expect(wrapper.find('h5').length).toEqual(1);
      expect(wrapper.find('form').length).toEqual(1);
      expect(wrapper).toMatchSnapshot();
    });
  });
  describe('calling methods directly from login page', () => {
    it('it should setstate of email when input email value is changed', () => {
      const event = {
        target: {
          name: 'email',
          value: 'teejay2k4@yahoo.com'
        }
      }
      expect(wrapper.state('email')).toBe('');
      wrapper.find(TextFieldGroup).at(0).simulate('change', event)
      expect(wrapper.state('email')).toBe('teejay2k4@yahoo.com');
    });
  });
  it('it should setstate of input password when input password value is changed', () => {
    const event = {
      target: {
        name: 'password',
        value: 'password'
      }
    }
    expect(wrapper.state('password')).toBe('');
    wrapper.find(TextFieldGroup).at(1).simulate('change', event)
    expect(wrapper.state('password')).toBe('password');
  });
  it('it should call the isValid function which should return true if there are no errors', () => {
    const instance = wrapper.instance()
    instance.setState({ email: 'teejay2k4@yahoo.com', password: 'testpassword'})
    expect (instance.isValid()).toEqual(true)
  });
  it('it should call the isValid function which should return false if there are errors', () => {
    const instance = wrapper.instance()
    instance.setState({ email: '', password: ''})
    expect (instance.isValid()).toEqual(false)
  });
});

