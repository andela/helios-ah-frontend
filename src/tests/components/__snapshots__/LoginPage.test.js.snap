// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Login Page render Login Page renders the login page component 1`] = `
<div
  className="container login-body"
>
  <div
    className="row"
  >
    <div
      className="col-sm-6"
    >
      <span
        className="login-image-container"
      />
    </div>
    <div
      className="col-lg-6 col-md-6 col-sm-6 col-xs-12 right-login"
    >
      <h3>
        WELCOME BACK.....
      </h3>
      <h5>
        LOGIN TO AUTHOR'S HAVEN
      </h5>
      <Connect(FlashMessageList)
        customAlertClass="custom-alert"
      />
      <form
        onSubmit={[Function]}
      >
        <TextFieldGroup
          errorFeedbackClass="login-invalid-feedback"
          field="email"
          onChange={[Function]}
          placeholder="Email Address"
          type="text"
          validFeedbackClass="login-valid-feedback"
          value=""
        />
        <TextFieldGroup
          errorFeedbackClass="login-invalid-feedback"
          field="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          validFeedbackClass="login-valid-feedback"
          value=""
        />
        <div
          className="forgot-password"
        >
          <a
            href="#"
          >
            <p>
              FORGOT PASSWORD?
            </p>
          </a>
        </div>
        <div
          className="login-buttons"
        >
          <SubmitButton
            buttonClass="login-btn"
            buttonValue="LOGIN"
            columnAttribute="login-col"
            isRequestSent={false}
          />
          <SubmitButton
            buttonClass="signup-btn"
            buttonValue="SIGNUP"
            columnAttribute="signup-col"
          />
        </div>
        <div
          className="social-media-icons"
        >
          <span
            className="facebook"
          />
          <span
            className="twitter"
          />
          <span
            className="google"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`Render Login Page render Login Page renders the login page component 2`] = `
<div
  className="container login-body"
>
  <div
    className="row"
  >
    <div
      className="col-sm-6"
    >
      <span
        className="login-image-container"
      />
    </div>
    <div
      className="col-lg-6 col-md-6 col-sm-6 col-xs-12 right-login"
    >
      <h3>
        WELCOME BACK.....
      </h3>
      <h5>
        LOGIN TO AUTHOR'S HAVEN
      </h5>
      <Connect(FlashMessageList)
        customAlertClass="custom-alert"
      />
      <form
        onSubmit={[Function]}
      >
        <TextFieldGroup
          errorFeedbackClass="login-invalid-feedback"
          field="email"
          onChange={[Function]}
          placeholder="Email Address"
          type="text"
          validFeedbackClass="login-valid-feedback"
          value=""
        />
        <TextFieldGroup
          errorFeedbackClass="login-invalid-feedback"
          field="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          validFeedbackClass="login-valid-feedback"
          value=""
        />
        <div
          className="forgot-password"
        >
          <a
            href="#"
          >
            <p>
              FORGOT PASSWORD?
            </p>
          </a>
        </div>
        <div
          className="login-buttons"
        >
          <SubmitButton
            buttonClass="login-btn"
            buttonValue="LOGIN"
            columnAttribute="login-col"
            isRequestSent={false}
          />
          <SubmitButton
            buttonClass="signup-btn"
            buttonValue="SIGNUP"
            columnAttribute="signup-col"
          />
        </div>
        <div
          className="social-media-icons"
        >
          <span
            className="facebook"
          />
          <span
            className="twitter"
          />
          <span
            className="google"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
